docs: config overview (#28035)
renovatebot/renovate: docs/usage/config-overview.md

# Renovate Configuration Overview

When Renovate runs on a repository, the final config in use is derived from:

- Global config
- Inherited config
- Repository config
- Resolved presets referenced in config

## Types of Config

> ### Global config

Reviewer: I think we should a reference that it is sometimes called  bot config. 

So people know that they can search the term in our discussions.

Author: Good point. I was even considering starting with a glossary of terms at the top of this page

Reviewer 2: If we can standardize on the terms, and use those consistently, we can add those terms to the [`abbreviations.md` file in the Renovate docs repository](https://github.com/renovatebot/renovatebot.github.io/blob/main/includes/abbreviations.md). That way when you read the docs online, you'll see a explanation when you hover over the term.

Also see the [Material for MkDocs manual, adding abbreviations and a glossary](https://squidfunk.github.io/mkdocs-material/reference/tooltips/#adding-abbreviations).

Or we can do both:

- Have glossary in this file directly
- Put the standardized terms in the `abbreviations.md` file

Author: I tried to use capitalization for "Global config", "Inherited config", "Onboarding config", etc. It would be good if these had hover explanations when viewed in docs

Author: Added bot config clarification 2 lines down

> Global config is the term for config defined by the person or team responsible for running the bot.

Author: ```suggestion
Global config is the term for config defined by the person or team responsible for running the bot.
Sometimes it's referred to as "bot config" because it's the config passed to the bot by the person running it.
```

Global config can contain config which is "global only" as well as any configuration options which are valid in Inherited config or Repository config.

If you are an end user of Renovate, such as a user of the Mend Renovate App, then you don't need to care too much about global config other than knowing that there are some settings you cannot change because they are global-only.
You can skip the rest of this "Global config" section and proceed to "Inherited config".

Global config can be read from file, environment, or CLI parameters.
You must configure at least one of these in order to provide your bot with the minimum details it needs to run, e.g. credentials.

#### File config

> The first place Renovate global config is read from is from file.

Reviewer: Maybe, it is worth to mention the default values here.

Reviewer 2: How do we ensure those stay up-to-date with the actual values? I'm always a bit wary of adding references to default values/behavior outside of the main "config options docs". Because the config option docs are likely to be updated when the code changes, and other mentions are easily forgotten.

Reviewer: I have been reffering it more in the direction of, these are the values first overwriting the default ones and then referencing the configuration options page

Author: Mention default values of what?

Reviewer: That gobal configs are applied over default values and where to find them.

Author: Added

By default Renovate checks for the existence of `config.js` in the current working directory, but you can override this by defining `RENOVATE_CONFIG_FILE` in env, e.g. `RENOVATE_CONFIG_FILE=/tmp/my-renovate-config.js`.

By default Renovate allows the config file to be _missing_ and does not error if it cannot find it, however if you have configured `RENOVATE_CONFIG_FILE` but the path you specified is not found then Renovate will error and exit, as it's assumed you have a configuration problem.
If the file is found but cannot be parsed then Renovate will also error and exit.

Global config files can be `.js` or `.json` files.
JS files can use synchronous or asyncronous methods inside, including even to fetch config information from remote hosts.

#### Environment config

> Global config can be defined using environment variables.

Reviewer: We probably should mention here that all env variables are prefixed with `RENOVATE_` and link exceptions like `LOG_LEVEL`, `LOG_CONTEXT` or the OTEL variables.

Author: Added

If the file is found but cannot be parsed then Renovate will also error and exit.

Global config files can be `.js` or `.json` files.
JS files can use synchronous or asyncronous methods inside, including even to fetch config information from remote hosts.

#### Environment config

Global config can be defined using environment variables.
Each config option which is supported in env has the prefix `RENOVATE_`.
For example, `RENOVATE_PLATFORM=gitlab` is the same as setting `"platform": "gitlab"` in File config.

Although the mapping from configuration option name to its corresponding Environment config name is fairly easy to understand, we recommend you consult the documentation for the field `env` for each option.
If the `env` field is missing then it means that the configuration option does not have its own Environment config variable name.

A special case for Environment config is the `RENOVATE_CONFIG` "meta" config option.

> The `RENOVATE_CONFIG` option accepts a stringified full config, e.g. `RENOVATE_CONFIG={"platform":"gitlab","onboarding":false}`.

Reviewer 3: I think we should add some samples for the different types, especially arrays are often a confusing issue when it comes to env vars.

Author: Added, see below

<!-- prettier-ignore -->
!!! warning
    Escaping punctionation can be challenging to get right in some environments, especially if you're passing stringified values.

Renovate additionally supports a list of "experimental" environment variables which start with `RENOVATE_X_` and are documented in [Self-hosted experimental environment variables](./self-hosted-experimental.md).
These variables are experimental, subject to change, and are not parsed as part of regular configuration.

Finally, there is a limited number of special environment variables which are loaded early prior to configuration parsing because they are used during logging initialization:

- `LOG_CONTEXT`: a unique identifier used in each log message to track context
- `LOG_FORMAT`: defaults to a "pretty" human-readable output, but can be changed to "json"
- `LOG_LEVEL`: most commonly used to change from the default `info` to `debug`

#### CLI config

The final way to configure Global config is through CLI parameters.
For example, the CLI parameter `--platform=gitlab` is the same as setting `"platform": "gitlab"` in File config or `RENOVATE_PLATFORM=gitlab` in Environment config.

> CLI config is read last and takes precedence over Environment and File config, so if you configure conflicting values in more than one of these then the one in CLI config will "win".

Reviewer 3: Or merged when option is mergable?

> CLI config is read last and takes precedence over Environment and File config, so if you configure conflicting values in more than one of these then the one in CLI config will "win".

Author: ```suggestion
CLI config is read last and takes precedence over Environment and File config, so if you configure conflicting values in more than one of these then the one in CLI config will be merged last and "win" if values conflict.
```

By default, Renovate will generate an "Onboarding PR" with a default config if a repository does not already have a Repository config file committed to the default branch.

### Onboarding Config

When Renovate creates an Onboarding PR it will propose a Repository config file to be merged.
By default, it is essentially an empty config with only the Renovate JSON schema referenced, but you can change this behavior if desired.

If you configure `onboardingConfig` in either Global config or Inherited config then Renovate will use that config directly instead of the default.

Alternatively if follow Renovate's naming convention for shared presets then it can automatically detect those instead.
If the repository `{{parentOrg}}/renovate-config` contains a `default.json` file then this will be treated as the organization's default preset and included in the Onboarding config.
Additionally for platforms which support nested Organization/Group hierarchies, Renovate will "hunt" up such hierarchies for a `renovate-config` repository with default config and stop when it finds the first.

<!-- prettier-ignore -->

> !!! note

Reviewer 3: ```suggestion
!!! warn
```

I think we should warn here? As we also warn in logs?

Author: I think note is enough. We have already warned about this in other places

Extending presets through Repository config is the most explicit and obvious way to use presets from an end user point of view.
If they wonder why behavior is a certain way, they can look at the Repository config and trace through all presets it references.

## Mend Renovate App Config

The [Mend Renovate App](https://github.com/apps/renovate) is a popular way to use Renovate on GitHub.com so it's important that any of its non-default behavior is documented here.

Importantly, logs for all Renovate jobs by the Mend Renovate App are available through the [Mend Developer Portal](https://developer.mend.io) so end users can view the logs to see which settings are applied.

### Onboarding behavior

If an Organization is installed with "All repositories" instead of "Selected repositories" then Renovate will default to "Silent" mode (`dryRun=lookup`).
This has been chosen for two reasons:

- Too often an account or org administrator selects this option without realizing it could cause the onboarding of hundreds of repositories, and

> - By offering this option, it means that org administrators _can_ install into All repositories without worrying about the noise, and then let individual Repository admins decide if/when to start onboarding

Reviewer 3: I think this currently can't be done by an repo admin. I was not able todo so for a repo in jenkinsci org.

Author: @nabeelsaabna can you check this point? We should allow repo admins to be able to override silent/interactive org defaults

Author: We'll treat this as a bug if it's not how it's working

By default, Renovate will create an "Onboarding PR" with a default config if a repository does not have a Repository config file committed to the default branch.

### Onboarding Config

When Renovate creates an Onboarding PR it will propose a Repository config file to be merged.
By default, it is essentially an empty config with only the Renovate JSON schema referenced, but you can change this behavior if desired.

If you configure `onboardingConfig` in either Global config or Inherited config then Renovate will use that config directly instead of the default.

Alternatively if you follow Renovate's naming convention for shared presets then it can automatically detect those instead.
If the repository `{{parentOrg}}/renovate-config` has a `default.json` file then this will be treated as the organization's default preset and included in the Onboarding config.
Additionally for platforms which support nested Organization/Group hierarchies, Renovate will "hunt" up such hierarchies for a `renovate-config` repository with default config and stop when it finds the first.

<!-- prettier-ignore -->
!!! note

>     Renovate will also check for `renovate.json` but this option is deprecated and is not recommended.

Author: ```suggestion
    Renovate will also check for `renovate.json` if it cannot find a `default.json`, however this option is deprecated and not recommended.
```

>     Renovate will also check for `renovate.json` but this option is deprecated and is not recommended.

Author: Does the below suggested edit from me make it clearer?

#### globalExtends

Sometimes you may not wish to put all settings within the Global config itself and instead commit it to a repository which is then referenced from the Global config.
In such cases, use `globalExtends` instead of `extends` so that it is resolved immediately and used as part of Global config.

#### extends

If you use `extends` within Global config then it's important to note that these are _not_ resolved/expanded during Global config processing and instead are passed through unresolved to be part of Repository config.
Passing `extends` through to be part of Repository config has two major consequences:

- It allows repository users to be able to use `ignorePresets` to ignore all or part of the `extends` presets, and
- Presets defined within `extends` in Global config will take _higher_ precedence that "regular" Global config, because it's resolved later

### Using a centralized config

Using "centralized" configs through Renovate presets is important in order to be able to:

- Save time by not repeating yourself in every repo with the same config, and
- Being able to change settings across an entire Organization or groups of repositories in one place

> Once you've created a centralized preset config, there are multiple ways you can pass it through to repositories:

Reviewer 2: I'm not a fan of having multiple ways to do things. I suppose this is needed for power users who "want to hide Renovate config" for whatever reason.

In any case maybe a table layout works better, where you add a column to rank the transparency. And a column that explains the gotcha's?

Author: We've grown so big that one size rarely fits all. It's not realistic to have just one way for things like this

- Defining it in Global config (either `globalExtends` or `extends`)
- Using it as your Inherited config, or referencing it from Inherited config using `extends`
- Ensuring it's referenced in Onboarding config so that it's committed as part of the Repository config

The above possibilities go from least to most transparent when it comes to end users.

Global config may be invisible to developers without log access, meaning they could be confused by any settings you apply - via presets or directly - within Global config.
For example the developers wonder why Renovate is behaving differently to its documented default behavior and may even think it's a bug.

Inherited config is visible to developers (it's within a repository they can see) although it's _implicitly_ applied so without log access and if they're not aware to look for an Inherited config repository then they may again be a little confused as to why default behavior has changed.

Extending presets through Repository config is the most explicit and obvious way to use presets from an end user point of view.

> If they wonder why behavior is a certain way, they can look at the Repository config and trace through all presets it references.

Reviewer 2: This section is hard to understand so I've stayed away from it.

I would prefer to have examples how to _avoid_ confusion. Right now it seems to explain how to create confusion.

Author: I've tried to improve this text with an edit

> Extending presets through Repository config is the most explicit and obvious way to use presets from an end user point of view.
> If they wonder why behavior is a certain way, they can look at the Repository config and trace through all presets it references.

Author: ```suggestion
The recommended approach for using a centralized preset is to explicitly "extend" it from every repository, which can be achieved easily if it's part of your `onboardingConfig`.
By having your centralized preset part of each Repository config `extends`, it has these two benefits:

- You still have the ability to change shared settings in a single location
- Any user viewing the repo can see the preset being extended and trace it back to understand which config is applied
```
